### PR TITLE
A place for discussion & code review

### PR DESCRIPTION
**Pull request by mingchaoliao**

### CODE REVIEW

**Comment by glennpai** [comment]
Tracking new feature requests unrelated to this review under -> https://github.com/glennpai/simple-graph-etl-python/discussions/2
Please continue to use this PR for review-related discussion.

File: simple_graph_etl/simpleetl.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        """
+        file_list = []
+
+        file_list_resp = requests.get(f'{self.library.base_url}/root:/{source_path}:/children',
+                                headers={'Authorization': 'Bearer ' + self.__token})
+
+        if file_list_resp.status_code == 200:
+            objs = file_list_resp.json()['value']
+
+            for obj in objs:
```
Comment (Maxxxxz) [inline simple_graph_etl/simpleetl.py:95]: I'd like to point these lines out in particular. There are a [few ways](https://docs.microsoft.com/en-us/graph/api/resources/driveitem?view=graph-rest-1.0#properties) to determine if an object is a file or a folder. You can check that it has the `file` property, and then use the download URL since you know it will exist. The second way is to just check that it has a download URL.
I think the way it is done here is fine since we don't care about folders, and only downloadable content in the drive. Just something worth pointing out since we can also use the properties to check for images and/or videos as well. This could be useful elsewhere.

Comment (glennpai) [reply simple_graph_etl/simpleetl.py:95]: Included change in pull -> https://github.com/glennpai/simple-graph-etl-python/pull/3
The existence of obj['file'] is almost the same as what we're doing above but this 'file' property will exist on all filetype objects. I don't know if there are any non-filetype objects that would have a downloadURL (folders do not) but this change should prevent us from grabbing those if there are.

File: simple_graph_etl/simpleetl.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        """
+        filenames = []
+        file_list_resp = requests.get(f'{self.library.base_url}/root:/{remote_path}:/children',
+                                headers={'Authorization': 'Bearer ' + self.__token})
+
+        if file_list_resp.status_code == 200:
+            objs = file_list_resp.json()['value']
+            for obj in objs:
+                if obj['file']:
+                    filenames.append(obj['name'])
```
Comment (mingchaoliao) [inline simple_graph_etl/simpleetl.py:95]: I think it can be simplified to:
```python
files = filter(lambda obj: obj['file'], objs)
filenames = [ file['name'] for file in files ]
```

File: simple_graph_etl/simpleetl.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                                headers={'Authorization': 'Bearer ' + self.__token})
+        if file_list_resp.status_code == 200:
+            objs = file_list_resp.json()['value']
+            for obj in objs:
+                if not obj['file']:
+                    continue
+                file_data = requests.get(obj['@microsoft.graph.downloadUrl'])
+                if file_data.status_code == 200:
+                    try:
+                        clean_path = re.sub(r'^(\\|\/)+|(\\|\/)+$', '', local_path)
```
Comment (mingchaoliao) [inline simple_graph_etl/simpleetl.py:122]: seems like you can use `os.path.normpath(...)` here https://docs.python.org/3/library/os.path.html#os.path.normpath

File: simple_graph_etl/simpleetl.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        and name
+
+        Parameters:
+            remote_path (string): Remote path of parent directory of file to delete
+            file_name (string): Name of remote file to delete
+        Returns:
+        """
+        list_url = f'{self.library.base_url}/root:/{remote_path}:/children'
+        file_list_response = requests.get(list_url,
+            headers={'Authorization': 'Bearer ' + self.__token})
```
Comment (mingchaoliao) [inline simple_graph_etl/simpleetl.py:149]: hmm, I saw similar code in the `filenames` method. To improve reusability, maybe you can create another method to handle the `list files/dirs` work and let `filenames` and `delete` methods use it rather than doing their own file listing.

File: simple_graph_etl/simpleetl.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+        if file_list_response.status_code == 200:
+            item_id = self.__get_item_id(file_list_response.json()['value'], file_name)
+            if item_id != '':
+                delete_url = f'{self.library.base_url}/items/'
+                delete_response = requests.delete(delete_url + item_id,
+                    headers={'Authorization': 'Bearer ' + self.__token})
+                if delete_response.status_code != 204:
+                    raise Exception(f'Failed to delete {file_name}. \
+                        {delete_response.raise_for_status()}')
```
Comment (mingchaoliao) [inline simple_graph_etl/simpleetl.py:159]: `raise_for_status` will actually raise an exception. It makes your `raise Exception(...)` pointless. There are several cases like that in this file.

**Review by mingchaoliao — COMMENTED** [review]
1. add a `requirements.txt` file containing your project's dependencies.
2. I found out that the file names are hard to follow as they are all lowercase and have no separator. I think most people prefer all lowercase and use underscores to improve readability. You can probably follow the PEP8 style guide https://peps.python.org/pep-0008/#package-and-module-names
3. **I encourage you to consider adding types to all possible places.** Similar to JavaScript, Python is a weak-type language. But writing strong-typed code (like why we replace JS with TypeScript) can greatly reduce type related bugs and improve code readability (as other developers can easily inspect the code in an IDE, e.g. know exactly the parameter structure, go to the class definition, etc.). see more about Python type hint system at https://docs.python.org/3/library/typing.html
4. It seems like Microsoft has a Graph API SDK for python, https://github.com/microsoftgraph/msgraph-sdk-python-core, are you able to utilize it? (instead of doing it manually using `requests`).
5. The way you download/upload files won't work well with large files. For large files (several Gig), people usually download/upload using streams, or do it in multiple chunks. Although I don't think we will run into that issue very often.

File: simple_graph_etl/simpleetl.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        Gets a list of file names that are children to the remote_path directory
+        Useful for checking existence of a remote file
+
+        Parameters:
+            remote_path (string): Path to parent directory containing target files
+        Returns:
+            filenames (string[]): List of file names in the remote_path directory
+        """
+        filenames = []
+        file_list_resp = requests.get(f'{self.library.base_url}/root:/{remote_path}:/children',
```
Comment (mingchaoliao) [inline simple_graph_etl/simpleetl.py:88]: You may be able to use `filter` and `select` query parameter to:
1. return files only https://docs.microsoft.com/en-us/graph/query-parameters#filter-parameter
2. return only the attributes you need  https://docs.microsoft.com/en-us/graph/query-parameters#select-parameter

This comment also applies to other places that make api calls.

**Comment by glennpai** [comment]
> 1. add a `requirements.txt` file containing your project's dependencies.
> 2. I found out that the file names are hard to follow as they are all lowercase and have no separator. I think most people prefer all lowercase and use underscores to improve readability. You can probably follow the PEP8 style guide https://peps.python.org/pep-0008/#package-and-module-names
> 3. **I encourage you to consider adding types to all possible places.** Similar to JavaScript, Python is a weak-type language. But writing strong-typed code (like why we replace JS with TypeScript) can greatly reduce type related bugs and improve code readability (as other developers can easily inspect the code in an IDE, e.g. know exactly the parameter structure, go to the class definition, etc.). see more about Python type hint system at https://docs.python.org/3/library/typing.html
> 4. It seems like Microsoft has a Graph API SDK for python, https://github.com/microsoftgraph/msgraph-sdk-python-core, are you able to utilize it? (instead of doing it manually using `requests`).
> 5. The way you download/upload files won't work well with large files. For large files (several Gig), people usually download/upload using streams, or do it in multiple chunks. Although I don't think we will run into that issue very often.

1. This can be done. 
2. This can be done. 
3. I will look into stronger typing. I prefer explicit type definitions but had not previously used them in Python. 
4. The official SDK is a powerful tool but unfortunately is not ideal for working with OIT SET restricted document libraries. From the authentication methods to our document library config, it is likely simpler to make a reusable request wrapper package for most actions than trying to weave in an existing SDK. 
5. This is an issue I would like to gauge how important it would be to fix. According to the Graph API docs, the current method [supports up to 60 MiB in one request](https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0#upload-bytes-to-the-upload-session). If we believe our SET integrations would ever need to upload files larger than 60 MiB, I would be open to modifying this to support larger files. It may also be worth creating a separate function to handle large uploads while leaving a simpler small-file upload function. I am open to ideas on how to approach this.
